### PR TITLE
Jormun: Manage cykleo

### PR DESCRIPTION
**Pull request by azime**

### CODE REVIEW

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    'url': 'www.cykleo.fr'
+}
+
+
+class CykleoError(RuntimeError):
+    pass
+
+
+class CykleoProvider(AbstractParkingPlacesProvider):
+    def __init__(self, url, network, service_id, username, password, operators={'cykleo'}, timeout=10,
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: default timeout should be at 2 sec

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+        self.username = username
+        self.password = password
+        self.operators = [o.lower() for o in operators]
+        self.timeout = timeout
+        fail_max = kwargs.get('circuit_breaker_max_fail', app.config['CIRCUIT_BREAKER_MAX_CYKLEO_FAIL'])
+        reset_timeout = kwargs.get('circuit_breaker_reset_timeout', app.config['CIRCUIT_BREAKER_CYKLEO_TIMEOUT_S'])
+        self.breaker = pybreaker.CircuitBreaker(fail_max=fail_max, reset_timeout=reset_timeout)
+        self._feed_publisher = FeedPublisher(**feed_publisher) if feed_publisher else None
+
+    @cache.memoize(app.config['CACHE_CONFIGURATION'].get('TIMEOUT_CYKLEO_JETON', 30))
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: default cache timeout for the token should be higher, at least 10 minutes

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            response = self.breaker.call(requests.post, '{}/pu/auth'.format(self.url),
+                                         timeout=self.timeout, headers=headers, data=json.dumps(data), verify=False)
+            if not response or response.status_code != 200:
+                raise CykleoError('cykleo, Invalid response')
+            content = response.json()
+            access_token = content.get("access_token")
+            if not access_token:
+                raise CykleoError('cykleo, access_token not exist in response')
+            return access_token
+        except Exception as e:
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: Error management can be improved

**Review by kinnou02 — CHANGES_REQUESTED** [review]

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        return None
+
+    @cache.memoize(app.config['CACHE_CONFIGURATION'].get('TIMEOUT_CYKLEO_JETON', 10*60))
+    def get_access_token(self):
+        headers = {
+            "Content-Type": "application/json",
+            "Accept": "application/json"
+        }
+        data = {"username": self.username, "password": self.password}
+        if self.service_id:
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: if service is `0` it won't be added to the dict, you need to explicitly check if it's `None`

**Comment by pbougue** [comment]
Jenkins, test this please.

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+        access_token = content.get("access_token")
+        if not access_token:
+            logging.getLogger(__name__).error('cykleo, access_token not exist in response')
+            return None
+        return access_token
+
+    @cache.memoize(app.config['CACHE_CONFIGURATION'].get('TIMEOUT_CYKLEO', 30))
+    def _call_webservice(self):
+        access_token = self.get_access_token()
+        headers = {'Authorization': '{}'.format(access_token)}
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: I don't see where you put the `Bearer` scheme

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py (hunk 4 above)
Comment (azime) [reply source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: :+1:

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py (hunk 5 above)
Comment (azime) [reply source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: Done

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py (hunk 4 above)
Comment (azime) [reply source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: Done

**Comment by azime** [comment]
Jenkins, test this please

**Review by kinnou02 — APPROVED** [review]

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+
+class CykleoProvider(AbstractParkingPlacesProvider):
+    def __init__(self, url, network, username, password, operators={'cykleo'}, certifi_verify=False,
+                 service_id=None, timeout=2,
+                 feed_publisher=DEFAULT_CYKLEO_FEED_PUBLISHER, **kwargs):
+        self.url = url
+        self.network = network.lower()
+        self.service_id = service_id
+        self.certifi_verify = certifi_verify
```
Comment (pbougue) [inline source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: I'd name it `verify_certificate`

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+            fail_max=kwargs.get('circuit_breaker_max_fail', app.config['CIRCUIT_BREAKER_MAX_CYKLEO_FAIL']),
+            reset_timeout=kwargs.get('circuit_breaker_reset_timeout', app.config['CIRCUIT_BREAKER_CYKLEO_TIMEOUT_S']))
+        self._feed_publisher = FeedPublisher(**feed_publisher) if feed_publisher else None
+
+    def service_caller(self, method, url, data=None, headers=None):
+        try:
+            kwargs = {"timeout": self.timeout, "verify": self.certifi_verify}
+            if headers:
+                kwargs.update({"headers": headers})
+            if data:
```
Comment (pbougue) [inline source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: Not sure the case can happen but I'd rather have `if data is not None` (same above for headers).

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+        return None
+
+    @cache.memoize(app.config['CACHE_CONFIGURATION'].get('TIMEOUT_CYKLEO_JETON', 10*60))
+    def get_access_token(self):
+        headers = {
+            "Content-Type": "application/json",
+            "Accept": "application/json"
+        }
+        data = {"username": self.username, "password": self.password}
+        if self.service_id is None:
```
Comment (pbougue) [inline source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: You probably forgot: is `not` None

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+            data.update({"serviceId": self.service_id})
+
+        response = self.service_caller(method=requests.post, url='{}/pu/auth'.format(self.url),
+                                       headers=headers, data=json.dumps(data))
+        if not response:
+            return None
+        content = response.json()
+        access_token = content.get("access_token")
+        if not access_token:
+            logging.getLogger(__name__).error('cykleo, access_token not exist in response')
```
Comment (pbougue) [inline source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: English: `cykleo, no access_token in response`

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+        ref = poi.get('properties', {}).get('ref')
+        data = self._call_webservice()
+        if not data:
+            return None
+        station = data.get(ref)
+        if not station:
+            return None
+        if 'availableDockCount' in station \
+                and ('availableClassicBikeCount' in station or 'availableElectricBikeCount' in station):
+            return Stands(station.get('availableDockCount'),
```
Comment (pbougue) [inline source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py:143]: Don't we also want to check that status is IN_SERVICE also? (poke @kinnou02 )

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py (hunk 8 above)
Comment (azime) [reply source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: Done

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py (hunk 9 above)
Comment (azime) [reply source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: Done

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py (hunk 7 above)
Comment (azime) [reply source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: In that case: if data is not None == if data

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py (hunk 6 above)
Comment (azime) [reply source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py]: Done

File: source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py (hunk 10 above)
Comment (azime) [reply source/jormungandr/jormungandr/parking_space_availability/bss/cykleo.py:143]: We do not know what to do with the status, for the moment we leave like that

**Pull request merged**